### ISSUE TITLE
KeyError: 'rabbitmq_plugin.enabled' while trying to install rabbitmq & setup a plugin at one single pass

### ISSUE THREAD

**Issue by pruiz** [saltstack/salt#27016]
Hi,

It looks like trying to install rabbitmq and then setting up some plugin/user/vhost/etc. against it inside the same highstate pass will fail miserably, even if 'reload_modules: True' is used.

This is a sample 'sample.sls' file which reproduces the issue using salt 2015.05.03 on CentOS-6.x.

```yaml
rabbitmq_plugin_rabbitmq_management:
  rabbitmq_plugin:
    - name: rabbitmq_management
    - enabled
    - runas: root
    - require:
      - pkg: rabbitmq-server
      - file: rabbitmq-env-link
      - file: rabbitmq-plugins-link
      - service: rabbitmq-server

rabbitmq-server:
  pkg:
    - latest
    - reload_modules: True
  service:
    - running
    - enable: True
    - require:
      - pkg: rabbitmq-server
    - watch:
      - pkg: rabbitmq-server

rabbitmq-env-link:
  file.symlink:
    - name: /usr/local/bin/rabbitmq-env
    - target: /usr/lib/rabbitmq/bin/rabbitmq-env
    - require:
      - pkg: rabbitmq-server

rabbitmq-plugins-link:
  file.symlink:
    - name: /usr/local/bin/rabbitmq-plugins
    - target: /usr/lib/rabbitmq/bin/rabbitmq-plugins
    - require:
      - pkg: rabbitmq-server
      - file: rabbitmq-env-link
```

Running this sls file using "salt-call -l debug state.template sample.sls" against a server _not_ having rabbitmq-server installed, throws the following output/error:

```
[DEBUG   ] Could not LazyLoad rabbitmq_plugin.enabled
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
KeyError: 'rabbitmq_plugin.enabled'
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 227, in salt_call
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/call.py", line 69, in run
    caller.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 236, in run
    ret = self.call()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 138, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 254, in template
    ret = st_.state.call_high(high_state)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 2067, in call_high
    ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1623, in call_chunks
    running = self.call_chunk(low, running, chunks)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1769, in call_chunk
    self._mod_init(low)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 612, in _mod_init
    self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
  File "/usr/lib/python2.6/site-packages/salt/utils/lazy.py", line 90, in __getitem__
    raise KeyError(key)
KeyError: 'rabbitmq_plugin.enabled'
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 227, in salt_call
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/call.py", line 69, in run
    caller.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 236, in run
    ret = self.call()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 138, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 254, in template
    ret = st_.state.call_high(high_state)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 2067, in call_high
    ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1623, in call_chunks
    running = self.call_chunk(low, running, chunks)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1769, in call_chunk
    self._mod_init(low)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 612, in _mod_init
    self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
  File "/usr/lib/python2.6/site-packages/salt/utils/lazy.py", line 90, in __getitem__
    raise KeyError(key)
KeyError: 'rabbitmq_plugin.enabled'
```

However, moving state 'rabbitmq_plugin_rabbitmq_management' to the end of sample.sls, worksaround the issue.. Something weird, as requirements should (IMHO) produce always the same resulting states ordering.

Here are the details from running environment:

```
[root@herma-test-01 ~]# cat /etc/redhat-release
CentOS release 6.3 (Final)

[root@herma-test-01 ~]# salt-call --versions-report
           Salt: 2015.5.3
         Python: 2.6.6 (r266:84292, Jul 23 2015, 15:22:56)
         Jinja2: 2.2.1
       M2Crypto: 0.20.2
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.0.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.5.0
           RAET: Not Installed
            ZMQ: 4.0.5
           Mako: Not Installed
        Tornado: Not Installed
```

**Comment by jfindlay** [comment]
@pruiz, thanks for the report.  Will you post the output of `salt-call -l debug state.show_sls sample.sls` for both layouts of `sample.sls` that you used?

**Comment by pruiz** [comment]
Hi,

state.show_sls assumes state to be at some env, instead of being a local file (as state.template), so I just got an error:

```
[root@herma-test-01 ~]# salt-call -l debug state.show_sls rabbitmq.sls
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/f_defaults.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/f_defaults.conf
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: herma-test-01.ono.ter0.com
[DEBUG   ] Configuration file path: /etc/salt/minion
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/f_defaults.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/f_defaults.conf
[DEBUG   ] The `dmidecode` binary is not available on the system. GPU grains will not be available.
[DEBUG   ] Initializing new SAuth for ('/etc/salt/pki/minion', 'herma-test-01.ono.ter0.com', 'tcp://172.21.201.224:4506')
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] LazyLoaded state.show_sls
[DEBUG   ] LazyLoaded saltutil.is_running
[DEBUG   ] Re-using SAuth for ('/etc/salt/pki/minion', 'herma-test-01.ono.ter0.com', 'tcp://172.21.201.224:4506')
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/f_defaults.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/f_defaults.conf
[DEBUG   ] The `dmidecode` binary is not available on the system. GPU grains will not be available.
[DEBUG   ] Re-using SAuth for ('/etc/salt/pki/minion', 'herma-test-01.ono.ter0.com', 'tcp://172.21.201.224:4506')
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[INFO    ] Loading fresh modules for state activity
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] Could not find file from saltenv 'base', 'salt://rabbitmq/sls.sls'
[DEBUG   ] Could not find file from saltenv 'base', 'salt://rabbitmq/sls/init.sls'
[DEBUG   ] Re-using SAuth for ('/etc/salt/pki/minion', 'herma-test-01.ono.ter0.com', 'tcp://172.21.201.224:4506')
[DEBUG   ] LazyLoaded nested.output
local:
    - No matching sls found for 'rabbitmq.sls' in env 'base'
```

**Comment by jfindlay** [comment]
Thanks for looking @pruiz.  I am going to try to replicate the issue.

**Comment by dferrantelli** [comment]
@jfindlay  I'm running into the same issue, let me know if there is anything i can do to help diagnose it.

**Comment by jfindlay** [comment]
@dferrantelli, your help is very appreciated.  Can you run `salt <minion> state.show_sls sample.sls` for the state @pruiz has posted, or your own state if you can also post the sls file as well?  Thanks.

**Comment by bsander** [comment]
I had exactly the same error in a similar setup, but it disappeared for me when I made sure a `rabbitmq` user/group existed beforehand:

```yaml
# Ensure service user and group exist; appears to fix nasty salt error
rabbitmq:
  group.present: []
  user.present:
    - gid_from_name: True
    - require:
      - group: rabbitmq
    - require_in:
      - pkg: rabbitmq-server
```

Not sure if it's a fix or a workaround though ;)

**Comment by jfindlay** [comment]
@bsander, that looks like a workaround to me.  Thanks for the information.

**Comment by jfindlay** [comment]
This may be similar to (part of) #26115.

**Comment by rallytime** [comment]
Hi All - This looks like the same regression/stacktrace as reported in #27619, #26882, #26454, and #26280 and should be fixed 2015.5.6 and 2015.5.0. Can anyone here test and confirm this either at one of those tags or at the HEAD of the 2015.8 or 2015.5 branch?

The backstory here is that this bug was fixed on the 2015.8 brach with #27681, which I back-ported to 2015.5 with #27681.

**Comment by dferrantelli** [comment]
Trying to install with git to the v2015.8.0 or v2015.8.1 tag in my vagrant setup results in the salt-master service not having an init script. and if i manually install them using the new repo, it reads most of my pillar items 3x, which causes naming conflicts. Installing 2015.5.6 through the repo does not have the pillar issue, but it doesn't succeed with rabbitmq either.

```
vagrant@app1-staging:~$ salt-call --version
salt-call 2015.5.6 (Lithium)
```
```
vagrant@salt-master1-staging:~$ salt-call --version
salt-call 2015.5.6 (Lithium)
vagrant@salt-master1-staging:~$ sudo salt app1staging state.highstate
app1staging:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1198, in _thread_return
        return_data = func(*args, **kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/modules/state.py", line 538, in highstate
        whitelist=kwargs.get('whitelist')
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2990, in call_highstate
        return self.state.call_high(high)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2068, in call_high
        ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1624, in call_chunks
        running = self.call_chunk(low, running, chunks)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1859, in call_chunk
        running = self.call_chunk(chunk, running, chunks)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1770, in call_chunk
        self._mod_init(low)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 613, in _mod_init
        self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
      File "/usr/lib/python2.7/dist-packages/salt/utils/lazy.py", line 87, in __getitem__
        return self._dict[key]
    KeyError: 'rabbitmq_plugin.enabled'
```

**Comment by rallytime** [comment]
Hrm. It appears I have misjudged when the back-porting fix went into the 2015.5 branch. I thought I got it in before we tagged 2015.5.6, but it appears that I didn't. I am seeing the same failure on the 2015.5.6 tag. 

@dferrantelli When using the HEAD of the 2015.5 branch, f72830700120a4f2c79eb6205075889ee4c29c47, however, that you tried above, I can't reproduce that stacktrace anymore.

I made a new salt-minion using Salt Cloud by installing from the 2015.5 branch:
```
root@rallytime:~# salt rally-u12-2015-5 test.versions_report
rally-u12-2015-5:
               Salt: 2015.5.6-81-gf728307
             Python: 2.7.3 (default, Feb 27 2014, 19:58:35)
             Jinja2: 2.6
           M2Crypto: 0.21.1
     msgpack-python: 0.1.10
       msgpack-pure: Not Installed
           pycrypto: 2.6.1
            libnacl: Not Installed
             PyYAML: 3.10
              ioflo: Not Installed
              PyZMQ: 14.5.0
               RAET: Not Installed
                ZMQ: 4.0.5
               Mako: Not Installed
            Tornado: Not Installed
            timelib: Not Installed
           dateutil: Not Installed
```
And then I ran the state provided above by @pruiz in a highstate call, and it worked perfectly:
```
# salt rally-u12-2015-5 state.highstate
rally-u12-2015-5:
----------
          ID: rabbitmq-server
    Function: pkg.latest
      Result: True
     Comment: The following packages were successfully installed/upgraded: rabbitmq-server
     Started: 09:50:09.774538
    Duration: 13404.231 ms
     Changes:
              ----------
              erlang-abi-13.a:
                  ----------
                  new:
                      1
                  old:
              erlang-asn1:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-base:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-corba:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-crypto:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-dev:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-diameter:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-docbuilder:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-edoc:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-erl-docgen:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-eunit:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-ic:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-inets:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-inviso:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-mnesia:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-nox:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-odbc:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-os-mon:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-parsetools:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-percept:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-public-key:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-runtime-tools:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-snmp:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-ssh:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-ssl:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-syntax-tools:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-tools:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-webtool:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              erlang-xmerl:
                  ----------
                  new:
                      1:14.b.4-dfsg-1ubuntu1
                  old:
              libltdl7:
                  ----------
                  new:
                      2.4.2-1ubuntu1
                  old:
              libodbc1:
                  ----------
                  new:
                      2.2.14p2-5ubuntu3
                  old:
              libsctp1:
                  ----------
                  new:
                      1.0.11+dfsg-2
                  old:
              lksctp-tools:
                  ----------
                  new:
                      1.0.11+dfsg-2
                  old:
              rabbitmq-server:
                  ----------
                  new:
                      2.7.1-0ubuntu4
                  old:
----------
          ID: rabbitmq-env-link
    Function: file.symlink
        Name: /usr/local/bin/rabbitmq-env
      Result: True
     Comment: Created new symlink /usr/local/bin/rabbitmq-env -> /usr/lib/rabbitmq/bin/rabbitmq-env
     Started: 09:50:23.181335
    Duration: 493.46 ms
     Changes:
              ----------
              new:
                  /usr/local/bin/rabbitmq-env
----------
          ID: rabbitmq-plugins-link
    Function: file.symlink
        Name: /usr/local/bin/rabbitmq-plugins
      Result: True
     Comment: Created new symlink /usr/local/bin/rabbitmq-plugins -> /usr/lib/rabbitmq/bin/rabbitmq-plugins
     Started: 09:50:23.681811
    Duration: 493.159 ms
     Changes:
              ----------
              new:
                  /usr/local/bin/rabbitmq-plugins
----------
          ID: rabbitmq-server
    Function: service.running
      Result: True
     Comment: Service restarted
     Started: 09:50:24.500095
    Duration: 1704.708 ms
     Changes:
              ----------
              rabbitmq-server:
                  True
----------
          ID: rabbitmq_plugin_rabbitmq_management
    Function: rabbitmq_plugin.enabled
        Name: rabbitmq_management
      Result: True
     Comment:
     Started: 09:50:26.206863
    Duration: 629.4 ms
     Changes:

Summary for rally-u12-2015-5
------------
Succeeded: 5 (changed=4)
Failed:    0
------------
Total states run:     5
Total run time:  16.725 s
```
For the sake of being complete, here's the versions report and the failure from another minion that is running 2015.5.6:
```
root@rallytime:~# salt rally-u12-2015-5-6 test.versions_report
rally-u12-2015-5-6:
               Salt: 2015.5.6
             Python: 2.7.3 (default, Feb 27 2014, 19:58:35)
             Jinja2: 2.6
           M2Crypto: 0.21.1
     msgpack-python: 0.1.10
       msgpack-pure: Not Installed
           pycrypto: 2.6.1
            libnacl: Not Installed
             PyYAML: 3.10
              ioflo: Not Installed
              PyZMQ: 14.5.0
               RAET: Not Installed
                ZMQ: 4.0.5
               Mako: Not Installed
            Tornado: Not Installed
            timelib: Not Installed
           dateutil: Not Installed
root@rallytime:~# salt rally-u12-2015-5-6 state.highstate
rally-u12-2015-5-6:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1198, in _thread_return
        return_data = func(*args, **kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/modules/state.py", line 538, in highstate
        whitelist=kwargs.get('whitelist')
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2990, in call_highstate
        return self.state.call_high(high)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2068, in call_high
        ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1624, in call_chunks
        running = self.call_chunk(low, running, chunks)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1770, in call_chunk
        self._mod_init(low)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 613, in _mod_init
        self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
      File "/usr/lib/python2.7/dist-packages/salt/utils/lazy.py", line 90, in __getitem__
        raise KeyError(key)
    KeyError: 'rabbitmq_plugin.enabled'
```

**Comment by dferrantelli** [comment]
@rallytime you are correct, human error on that last test. It is working on f728307.

**Comment by rallytime** [comment]
Sweet! With your confirmation, and with me testing the same states that were provided by @pruiz, I'm going to go a head and close this. This fix will be available in the next release of the 2015.5 branch, 2015.5.7.
